lib/ReportFunctions.groovy

```diff
@@ -0,0 +1,160 @@
+
+
+
+
+class ReportFunctions{
+   
+/*
+    enum FuncType{
+        GENERIC("GENERIC"),
+        CONTIG("CONTIG"),
+        READQUALITY("READQUALITY"),
+        AUTOFAIL("AUTOFAIL"),
+        COVERAGE("COVERAGE")
+         
+    }
+*/
+
+
+    enum FuncType{
+        GENERIC,
+        CONTIG,
+        READQUALITY,
+        AUTOFAIL,
+        COVERAGE 
+    }
+
+
+    static def select_qc_func(java.util.LinkedHashMap qual_data, java.lang.String metric, java.util.ArrayList qc_message, java.util.LinkedHashMap meta_info, java.lang.String func){
+        def check_failed = 0
+        def reisolate = 0
+        def resequence = 0
+        def check_ignored = 0
+        def failed_p = false
+        def checks = 0
+        def function = func as FuncType
+
+        switch(function){
+            case FuncType.GENERIC:
+                (checks,
+                reisolate, 
+                resequence, 
+                failed_p, 
+                check_failed, 
+                check_ignored) = this.generic_qc_func(qual_data, metric, qc_message)
+                break;
+            case FuncType.AUTOFAIL:
+                (checks,
+                reisolate, 
+                resequence, 
+                failed_p, 
+                check_failed, 
+                check_ignored) = this.generic_qc_func(qual_data, metric, qc_message)
+                break;
+            case FuncType.READQUALITY:
+                if(!meta_info.assembly){ 
+                    (checks,
+                    reisolate, 
+                    resequence, 
+                    failed_p, 
+                    check_failed, 
+                    check_ignored) = this.generic_qc_func(qual_data, metric, qc_message)
+                }
+                break;
+            case FuncType.COVERAGE:
+                if(!meta_info.assembly){ 
+                    (checks,
+                    reisolate, 
+                    resequence, 
+                    failed_p, 
+                    check_failed, 
+                    check_ignored) = this.generic_qc_func(qual_data, metric, qc_message)
+                    if(!failed_p && meta_info.downsampled){
+                        qc_message.add("The sample may have been downsampled too aggressively, if this is the cause please re-run sample with a different target depth.")
+                    }
+                }
+                break;
+            case FuncType.CONTIG:
+                (checks,
+                reisolate, 
+                resequence, 
+                failed_p, 
+                check_failed, 
+                check_ignored) = this.contig_qc_func(qual_data, metric, qc_message)
+                break;
+        }
```

Comment:
Could you add a `default` case here that raises an exception about an invalid FuncType (though I suspect you would likely also get an exception raised when casting the string to the enum FuncType anyways).